File: sirenhead_tool/meta.py
"""Project metadata (`.meta.json`) read/write.

Per docs/shell-layer/13-meta-json.md, `.meta.json` is the
**derived project metadata** — the CLI's bookkeeping file that
records the actual mapping between project files and original
game resources. It is NOT a copy of `gallate.yaml`.

Atomic write pattern:

  1. Build the new metadata object in memory.
  2. Write to a temporary file in the same directory.
  3. Validate that the temporary file parses back.
  4. `os.replace` the temp file onto `.meta.json` (atomic on POSIX
     and Windows when both paths are on the same filesystem).

If any step fails, the existing `.meta.json` (if any) is left
untouched.
"""

from __future__ import annotations

import json
import os
import re
import tempfile
from datetime import datetime, timezone
from pathlib import Path

from .discovery import CLI_ID, CLI_VERSION
from .renpy_extract import ExtractedString

META_SCHEMA_VERSION = 1
"""Bump when the on-disk format changes incompatibly.

Per docs/shell-layer/13 § 13.4.1, the CLI refuses to load a
`.meta.json` whose `schema_version` is higher than the CLI knows.
"""

# Engine id is reported in `.meta.json` per docs/shell-layer/13 §
# 13.4.1. We use the same engine id as in `manifest.targets`.
ENGINE_ID = "renpy"


def now_iso() -> str:
    """ISO-8601 UTC timestamp with second precision and Z suffix.

    Ren'Py's bundled tooling prefers this exact shape.
    """
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def sha256_file(path: Path) -> str:
    """Compute the lowercase hex SHA-256 of a file's bytes."""
    import hashlib

    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return f"sha256:{h.hexdigest()}"


def new_meta(
    project_root: Path,
    entries: list[dict],
    schema_version: int = META_SCHEMA_VERSION,
) -> dict:
    """Build a fresh `.meta.json` object from a list of file entries.

    Each entry has the shape from `extract.build_meta_entries`.
    Returns a dict ready for `atomic_write`.
    """
    return {
        "schema_version": schema_version,
        "generated_by": {"id": CLI_ID, "version": CLI_VERSION},
        "generated_at": now_iso(),
        "project": {"root": "."},
        "files": entries,
    }


def read_meta(project_root: Path) -> dict | None:
    """Load `.meta.json` from the project root.

    Returns None if the file doesn't exist.
    Raises ValueError if the schema_version is higher than this
    CLI knows (per docs/shell-layer/13 § 13.8).
    """
    path = project_root / ".meta.json"
    if not path.is_file():
        return None
    data = json.loads(path.read_text(encoding="utf-8"))
    sv = data.get("schema_version")
    if not isinstance(sv, int):
        raise ValueError(f".meta.json has non-integer schema_version: {sv!r}")
    if sv > META_SCHEMA_VERSION:
        raise ValueError(
            f".meta.json schema_version {sv} is higher than this CLI "
            f"knows ({META_SCHEMA_VERSION}). Refusing to load."
        )
    return data


def atomic_write(project_root: Path, meta: dict) -> None:
    """Write `.meta.json` atomically.

    Pattern: temp file → validate → `os.replace`. The temp file
    lives in the same directory as the final file so `os.replace`
    is atomic on the same filesystem.
    """
    final = project_root / ".meta.json"
    fd, tmp_path = tempfile.mkstemp(
        prefix=".meta.", suffix=".json.tmp", dir=str(project_root),
    )
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            json.dump(meta, f, ensure_ascii=False, indent=2)
            f.write("\n")
        # Validate the temp file parses back identically.
        with open(tmp_path, encoding="utf-8") as f:
            reread = json.load(f)
        if reread != meta:
            raise RuntimeError(
                ".meta.json round-trip failed: written != re-read"
            )
        os.replace(tmp_path, final)
    except Exception:
        # Best-effort cleanup of the temp file.
        try:
            os.unlink(tmp_path)
        except OSError:
            pass
        raise


def build_meta_entries(
    project_root: Path,
    units_by_file: dict[Path, list[dict]],
    source_root: Path,
) -> list[dict]:
    """Build the `files[]` entries from a set of unit-file results.

    `units_by_file` maps `Path` (project file under
    `text/units/`) to a list of unit dicts. We pair each project
    file with its source .rpy file (looked up via `extensions`),
    the CLI/engine ids, and the units' aggregate size + hash.

    The mapping from project file → source file is encoded in the
    unit entries themselves (the CLI emits a `source_file` field
    per unit). We extract the unique source files here.
    """
    entries: list[dict] = []
    for project_file, units in units_by_file.items():
        if not units:
            continue
        # All units in this project file share the same source.
        source_rel = units[0]["source_file"]
        source_abs = source_root / source_rel
        size = project_file.stat().st_size if project_file.is_file() else 0
        # Use the bytes we already have in memory for the hash to
        # avoid a second read.
        h = sha256_file(project_file) if project_file.is_file() else None
        entries.append({
            "project": str(project_file.relative_to(project_root)),
            "source": source_rel,
            "type": "text",
            "cli": {"id": CLI_ID, "version": CLI_VERSION},
            "engine": {"id": ENGINE_ID},
            "sub_media": units[0].get("sub_media", "dialog"),
            "size": size,
            "hash": h,
            "extensions": {
                CLI_ID: {
                    "renpy_kind": units[0].get("renpy_kind", "dialog"),
                    "byte_count": sum(
                        len(u["source"]) for u in units
                    ),
                },
            },
        })
    return entries

File: internal/sirenhead/meta.go
package sirenhead

// `.meta.json` — derived project metadata (docs/shell-layer/13-meta-json.md).
//
// `.meta.json` is the CLI's bookkeeping file: it records the actual
// mapping between gallate project files and original game resources.
// It is NOT a copy of `gallate.yaml` (intent) and NOT the CLI manifest
// (identity).
//
// Two rules matter here:
//
//   - Ownership (§ 13.10): entries written by another CLI are never
//     silently overwritten; this CLI reads the existing file, keeps
//     foreign `files[]` entries verbatim, and only replaces its own.
//   - Atomicity (§ 13.9): temp file → validate by reading back → rename.

import (
	"encoding/json"
	"os"
	"path/filepath"
)

// MetaSchemaVersion is the `.meta.json` format version this CLI writes
// and the highest it will load.
const MetaSchemaVersion = 1

// MetaEntry is one `files[]` mapping entry.
type MetaEntry struct {
	Project    string
	Source     string
	Type       string
	SubMedia   string
	Size       int64
	Hash       string
	ResourceID string
	Encoding   string
	CLI        struct{ ID, Version string }
	Engine     struct{ ID, Version string }
	Raw        map[string]any // the entry as read, for verbatim preservation
}

// MetaDoc is a loaded `.meta.json`.
type MetaDoc struct {
	SchemaVersion int
	GeneratedAt   string
	Root          string
	Files         []MetaEntry
}

// readMeta loads `.meta.json` from the project root.
//
// A missing file returns (nil, nil). A `schema_version` higher than
// this CLI knows is refused (§ 13.4.1) rather than misinterpreted.
func readMeta(projectRoot string) (*MetaDoc, error) {
	path := filepath.Join(projectRoot, ".meta.json")
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, exitErr(ExitProjectConfig, "cannot read .meta.json: "+err.Error())
	}
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, exitErr(ExitProjectConfig, ".meta.json is corrupted: "+err.Error())
	}
	doc := &MetaDoc{}
	if v, ok := asInt(raw["schema_version"]); ok {
		doc.SchemaVersion = v
	} else {
		return nil, exitErr(ExitProjectConfig, ".meta.json has a non-integer schema_version")
	}
	if doc.SchemaVersion > MetaSchemaVersion {
		return nil, exitErr(ExitProjectConfig,
			"`.meta.json` schema_version is higher than this CLI knows; refusing to load")
	}
	doc.GeneratedAt, _ = raw["generated_at"].(string)
	if proj := asMap(raw["project"]); proj != nil {
		doc.Root, _ = proj["root"].(string)
	}
	entries, _ := raw["files"].([]any)
	for _, item := range entries {
		m, ok := item.(map[string]any)
		if !ok {
			continue
		}
		e := MetaEntry{Raw: m}
		e.Project, _ = m["project"].(string)
		e.Source, _ = m["source"].(string)
		e.Type, _ = m["type"].(string)
		e.SubMedia, _ = m["sub_media"].(string)
		e.Hash, _ = m["hash"].(string)
		e.ResourceID, _ = m["resource_id"].(string)
		e.Encoding, _ = m["encoding"].(string)
		if n, ok := asInt(m["size"]); ok {
			e.Size = int64(n)
		}
		if cli := asMap(m["cli"]); cli != nil {
			e.CLI.ID, _ = cli["id"].(string)
			e.CLI.Version, _ = cli["version"].(string)
		}
		if eng := asMap(m["engine"]); eng != nil {
			e.Engine.ID, _ = eng["id"].(string)
			e.Engine.Version, _ = eng["version"].(string)
		}
		doc.Files = append(doc.Files, e)
	}
	return doc, nil
}

// foreignMetaEntries returns the entries of an existing `.meta.json`
// that belong to another CLI, in their original shape.
func foreignMetaEntries(existing *MetaDoc) []any {
	var out []any
	if existing == nil {
		return nil
	}
	for _, e := range existing.Files {
		if e.CLI.ID != "" && e.CLI.ID != CLIID {
			out = append(out, e.Raw)
		}
	}
	return out
}

// textMetaEntry builds a `files[]` entry for one text unit file, in the
// reference implementation's key order.
func textMetaEntry(projectRoot string, w writtenUnitFile) (*orderedMap, error) {
	abs := filepath.Join(projectRoot, filepath.FromSlash(w.rel))
	hash, err := sha256File(abs)
	if err != nil {
		return nil, exitErr(ExitOutputFailure, err.Error())
	}
	var size int64
	if st, serr := os.Stat(abs); serr == nil {
		size = st.Size()
	}
	// docs/shell-layer/13 § 13.4.1: `sub_media` describes the resource's
	// sub-media category. The reference implementation records the first
	// unit's kind here; a unit file can of course mix kinds.
	subMedia := ""
	byteCount := 0
	if len(w.units) > 0 {
		subMedia = w.units[0].Kind
	}
	for _, u := range w.units {
		byteCount += len(u.Source) // UTF-8 bytes, as the key name says
	}
	m := &orderedMap{}
	m.Set("project", w.rel)
	if len(w.sources) == 1 {
		m.Set("source", w.sources[0])
	}
	m.Set("type", "text")
	cli := &orderedMap{}
	cli.Set("id", CLIID)
	cli.Set("version", CLIVersion)
	m.Set("cli", cli)
	eng := &orderedMap{}
	eng.Set("id", EngineID)
	m.Set("engine", eng)
	m.Set("sub_media", subMedia)
	m.Set("size", size)
	m.Set("hash", hash)
	ext := &orderedMap{}
	ours := &orderedMap{}
	ours.Set("renpy_kind", subMedia)
	ours.Set("byte_count", byteCount)
	ext.Set(CLIID, ours)
	m.Set("extensions", ext)
	return m, nil
}

// newMetaDoc assembles a `.meta.json` document: this CLI's entries plus
// any foreign entries preserved from the previous file.
func newMetaDoc(entries []any) *orderedMap {
	doc := &orderedMap{}
	doc.Set("schema_version", MetaSchemaVersion)
	gen := &orderedMap{}
	gen.Set("id", CLIID)
	gen.Set("version", CLIVersion)
	doc.Set("generated_by", gen)
	doc.Set("generated_at", nowISO())
	proj := &orderedMap{}
	proj.Set("root", ".")
	doc.Set("project", proj)
	doc.Set("files", entries)
	return doc
}

// writeMeta writes `.meta.json` atomically.
func writeMeta(projectRoot string, doc *orderedMap) error {
	path := filepath.Join(projectRoot, ".meta.json")
	if err := atomicWriteJSON(path, doc); err != nil {
		return exitErr(ExitOutputFailure, "cannot write .meta.json: "+err.Error())
	}
	return nil
}
